tools/module_diag_hailcast.F90

```diff
@@ -0,0 +1,1870 @@
+       !   RWA_new(k) = RWA_adiabat(k)*(h1d(k)-h1d(k-1)) - RWA_new(k-1)
+       !   IF (RWA_new(k).LT.0) RWA_new(k) = 0.
+       !ENDIF
+    ENDDO
+    !remove the height factor from RWA_new
+    DO k=KBAS,nz
+       RWA_new(k) = RWA_new(k) / (h1d(k)-h1d(k-1))
+    ENDDO
+
+
+
+    !!!!!!!!!!!!!!!! 3. INITIAL EMBRYO SIZE  !!!!!!!!!!!!!!!!!!!!!
+    !!!      SET CONSTANT RANGE OF INITIAL EMBRYO SIZES        !!!
+    !!!                     START LOOP                         !!!
+    !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+    ! See Adams-Selin and Ziegler 2016 MWR for explanation of why
+    ! these sizes were picked.
+    !Run each initial embryo size perturbation
+    DO i=1,5
+      SELECT CASE (i)
+        CASE (1)
+        !Initial hail embryo diameter in m, at cloud base
+          DD = 5.E-3
+          tk_embryo = 265.155  !-8C
+        CASE (2)
+          DD = 7.5E-3
+          tk_embryo = 265.155  !-8C
+        CASE (3)
+          DD = 5.E-3
+          tk_embryo = 260.155  !-13C
+        CASE (4)
+          DD = 7.5E-3
+          tk_embryo = 260.155  !-13C
+        CASE (5)
+          tk_embryo = 260.155  !-13C
+          DD = 1.E-2
+      END SELECT
+
+
+      RTIME = 2000.
+      IF (wdur .LT. RTIME) RTIME = wdur
+
+      TFZL = tk_embryo
+      CALL INTERPP(PA, WFZLP, TCA, tk_embryo, IFOUT, nz)
+      CALL INTERP(h1d, ZFZL, WFZLP, IFOUT, PA, nz)
+      CALL INTERP(RA,  RFZL, WFZLP, IFOUT, PA, nz)
+      CALL INTERP(VUU_pert, VUFZL, WFZLP, IFOUT, PA, nz)
+      CALL INTERP(rho1d, DENSAFZL, WFZLP, IFOUT, PA, nz)
+
+      !Begin hail simulation time (seconds)
+      sec = 0.
+
+
+      !!!!!!!!!!!!!!!!!!  4. INITIAL PARAMETERS    !!!!!!!!!!!!!!!!!
+      !!!          PUT INITIAL PARAMETERS IN VARIABLES           !!!
+      !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+      !Set initial values for parameters at freezing level
+      P = WFZLP
+      RS = RFZL
+      TC = TFZL
+      VU = VUFZL
+      Z = ZFZL - ht
+      LDEPTH = Z
+      DENSA = DENSAFZL
+
+      !Set initial hailstone parameters
+      nofroze=1 !Set test for embryo: 0 for never been frozen; 1 frozen
+      TS = TC
+      TSm1 = TS
+      TSm2 = TS
+      D = DD   !hailstone diameter in m
+      FW = 0.0
+      DENSE = 500.  !kg/m3
+      ITYPE=1.  !Assume starts in dry growth.
+      CLOUDON=1  !we'll eventually turn cloud "off" once updraft past time limit
+
+      !Start time loop.
+      DO WHILE (sec .lt. TAU)
+         sec = sec + secdel
+
+         !!!!!!!!!!!!!!!!!!  5. CALCULATE PARAMETERS  !!!!!!!!!!!!!!!!!
+         !!!              CALCULATE UPDRAFT PROPERTIES              !!!
+         !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+         !Intepolate vertical velocity to our new pressure level
+         CALL INTERP(VUU_pert,VUMAX,P,IFOUT,PA,nz)
+         !print *, 'INTERP VU: ', VU, P
+
+         !Outside pressure levels?  If so, exit loop
+         IF (IFOUT.EQ.1) GOTO 100
```

Comment:
Please rewrite the code to remove all the GOTO statements, it is not allowed in operational model code.